filters/builtin/compress.go

```diff
@@ -316,7 +318,7 @@ func encode(out *io.PipeWriter, in io.ReadCloser, enc string, level int) {
 		pe := pool.Get()
 		if pe != nil {
 			e = pe.(encoder)
-			defer pool.Put(pe)
+			backToPool = func() { pool.Put(pe) }
```

Comment:
This whole function is rather complicated to follow. I'd simplify it a bit:
 * Drop the `backToPool` variable.
 * Move the `level` check and putting the encoder to the main closing defer. It's also a good idea to check the error from `e.Close()` and refuse to pool the encoder back.
 * Use explicit `if/else` statements instead of re-checking for `if e == nil`.
 * `pool.Get()` can't return a `nil` if used with `p.New`, so just drop this thing altogether (or simply panic).

Reply:
It might also be a good idea to `Reset` _before_ putting things back into the pool instead of the other way around.